File: backend/src/main/java/com/leadgrowth/controller/UserController.java
package com.leadgrowth.controller;

import com.leadgrowth.dto.PasswordChangeRequest;
import com.leadgrowth.dto.UserProfileRequest;
import com.leadgrowth.dto.UserInviteRequest;
import com.leadgrowth.dto.UserRoleUpdateRequest;
import com.leadgrowth.dto.UserStatusUpdateRequest;
import com.leadgrowth.entity.Invitation;
import com.leadgrowth.entity.User;
import com.leadgrowth.service.UserService;
import jakarta.validation.Valid;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/api/users")
public class UserController {

    private final UserService userService;
    private final com.leadgrowth.service.ProductivityService productivityService;

    public UserController(UserService userService, com.leadgrowth.service.ProductivityService productivityService) {
        this.userService = userService;
        this.productivityService = productivityService;
    }

    @PutMapping("/profile")
    public ResponseEntity<User> updateProfile(@Valid @RequestBody UserProfileRequest request) {
        String email = SecurityContextHolder.getContext().getAuthentication().getName();
        return ResponseEntity.ok(userService.updateProfile(request, email));
    }

    @PostMapping("/password")
    public ResponseEntity<Void> changePassword(@Valid @RequestBody PasswordChangeRequest request) {
        String email = SecurityContextHolder.getContext().getAuthentication().getName();
        userService.changePassword(request, email);
        return ResponseEntity.ok().build();
    }

    @GetMapping("/members")
    public ResponseEntity<List<User>> getWorkspaceMembers() {
        String email = SecurityContextHolder.getContext().getAuthentication().getName();
        return ResponseEntity.ok(userService.getWorkspaceMembers(email));
    }

    @PostMapping("/invite")
    public ResponseEntity<Invitation> inviteUser(@Valid @RequestBody UserInviteRequest request) {
        String email = SecurityContextHolder.getContext().getAuthentication().getName();
        return ResponseEntity.ok(userService.inviteUser(request, email));
    }

    @PutMapping("/{id}/role")
    public ResponseEntity<User> updateUserRole(@PathVariable Long id, @Valid @RequestBody UserRoleUpdateRequest request) {
        String email = SecurityContextHolder.getContext().getAuthentication().getName();
        return ResponseEntity.ok(userService.updateUserRole(id, request.getRole(), email));
    }

    @PutMapping("/{id}/status")
    public ResponseEntity<User> updateUserStatus(@PathVariable Long id, @Valid @RequestBody UserStatusUpdateRequest request) {
        String email = SecurityContextHolder.getContext().getAuthentication().getName();
        return ResponseEntity.ok(userService.updateUserStatus(id, request.getStatus(), email));
    }

    @PutMapping("/{id}/details")
    public ResponseEntity<User> editUserDetails(@PathVariable Long id, @Valid @RequestBody UserProfileRequest request) {
        String email = SecurityContextHolder.getContext().getAuthentication().getName();
        return ResponseEntity.ok(userService.editUserDetails(id, request, email));
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<Void> deleteUser(@PathVariable Long id) {
        String email = SecurityContextHolder.getContext().getAuthentication().getName();
        userService.deleteUser(id, email);
        return ResponseEntity.ok().build();
    }

    @PostMapping("/{id}/reset-password")
    public ResponseEntity<User> resetUserPassword(@PathVariable Long id, @RequestParam String newPassword) {
        String email = SecurityContextHolder.getContext().getAuthentication().getName();
        return ResponseEntity.ok(userService.resetUserPassword(id, newPassword, email));
    }

    @PostMapping("/transfer-ownership")
    public ResponseEntity<com.leadgrowth.entity.Workspace> transferWorkspaceOwnership(@RequestParam Long newOwnerId) {
        String email = SecurityContextHolder.getContext().getAuthentication().getName();
        return ResponseEntity.ok(userService.transferWorkspaceOwnership(newOwnerId, email));
    }

    @PutMapping("/availability")
    public ResponseEntity<User> updateAvailabilityStatus(@RequestParam String status) {
        String email = SecurityContextHolder.getContext().getAuthentication().getName();
        return ResponseEntity.ok(userService.updateAvailabilityStatus(status, email));
    }

    @GetMapping("/productivity")
    public ResponseEntity<List<com.leadgrowth.dto.TeamProductivityDto>> getTeamProductivity() {
        String email = SecurityContextHolder.getContext().getAuthentication().getName();
        return ResponseEntity.ok(productivityService.getTeamProductivity(email));
    }
}

File: backend-dotnet/Controllers/UserController.cs
using System.Security.Claims;
using LeadGrowth.DTOs;
using LeadGrowth.Models;
using LeadGrowth.Services;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace LeadGrowth.Controllers;

[ApiController]
[Route("api/users")]
[Authorize]
public class UserController : ControllerBase
{
    private readonly IUserService _userService;
    private readonly IProductivityService _productivityService;

    public UserController(IUserService userService, IProductivityService productivityService)
    {
        _userService = userService;
        _productivityService = productivityService;
    }

    [HttpPut("profile")]
    public async Task<ActionResult<User>> UpdateProfile([FromBody] UserProfileRequest request)
    {
        var email = GetUserEmail();
        var user = await _userService.UpdateProfileAsync(request, email);
        return Ok(user);
    }

    [HttpPost("change-password")]
    public async Task<IActionResult> ChangePassword([FromBody] PasswordChangeRequest request)
    {
        var email = GetUserEmail();
        try
        {
            await _userService.ChangePasswordAsync(request, email);
            return Ok(new { message = "Password updated successfully" });
        }
        catch (ArgumentException ex)
        {
            return BadRequest(new { message = ex.Message });
        }
    }

    [HttpGet("members")]
    public async Task<ActionResult<List<User>>> GetWorkspaceMembers()
    {
        var email = GetUserEmail();
        try
        {
            var members = await _userService.GetWorkspaceMembersAsync(email);
            return Ok(members);
        }
        catch (InvalidOperationException ex)
        {
            return BadRequest(new { message = ex.Message });
        }
    }

    [HttpGet("assignable")]
    public async Task<ActionResult<List<User>>> GetAssignableUsers()
    {
        var email = GetUserEmail();
        try
        {
            var assignable = await _userService.GetAssignableUsersAsync(email);
            return Ok(assignable);
        }
        catch (InvalidOperationException ex)
        {
            return BadRequest(new { message = ex.Message });
        }
    }

    [HttpPost("invite")]
    public async Task<ActionResult<Invitation>> InviteUser([FromBody] UserInviteRequest request)
    {
        var email = GetUserEmail();
        try
        {
            var invitation = await _userService.InviteUserAsync(request, email);
            return Ok(invitation);
        }
        catch (ArgumentException ex)
        {
            return BadRequest(new { message = ex.Message });
        }
        catch (InvalidOperationException ex)
        {
            return BadRequest(new { message = ex.Message });
        }
    }

    [HttpPut("{userId}/role")]
    [Authorize(Policy = "RequireAdmin")]
    public async Task<ActionResult<User>> UpdateUserRole(long userId, [FromBody] UserRoleUpdateRequest request)
    {
        var email = GetUserEmail();
        try
        {
            var user = await _userService.UpdateUserRoleAsync(userId, request.Role, email);
            return Ok(user);
        }
        catch (ArgumentException ex)
        {
            return BadRequest(new { message = ex.Message });
        }
        catch (InvalidOperationException ex)
        {
            return BadRequest(new { message = ex.Message });
        }
    }

    [HttpPut("{userId}/status")]
    [Authorize(Policy = "RequireManagerOrAdmin")]
    public async Task<ActionResult<User>> UpdateUserStatus(long userId, [FromBody] UserStatusUpdateRequest request)
    {
        var email = GetUserEmail();
        try
        {
            var user = await _userService.UpdateUserStatusAsync(userId, request.Status, email);
            return Ok(user);
        }
        catch (ArgumentException ex)
        {
            return BadRequest(new { message = ex.Message });
        }
        catch (InvalidOperationException ex)
        {
            return BadRequest(new { message = ex.Message });
        }
    }

    [HttpPut("{userId}/details")]
    [Authorize(Policy = "RequireManagerOrAdmin")]
    public async Task<ActionResult<User>> EditUserDetails(long userId, [FromBody] UserProfileRequest request)
    {
        var email = GetUserEmail();
        try
        {
            var user = await _userService.EditUserDetailsAsync(userId, request, email);
            return Ok(user);
        }
        catch (ArgumentException ex)
        {
            return BadRequest(new { message = ex.Message });
        }
        catch (InvalidOperationException ex)
        {
            return BadRequest(new { message = ex.Message });
        }
    }

    [HttpDelete("{userId}")]
    [Authorize(Policy = "RequireAdmin")]
    public async Task<IActionResult> DeleteUser(long userId)
    {
        var email = GetUserEmail();
        try
        {
            await _userService.DeleteUserAsync(userId, email);
            return Ok(new { message = "User deleted successfully" });
        }
        catch (ArgumentException ex)
        {
            return BadRequest(new { message = ex.Message });
        }
        catch (InvalidOperationException ex)
        {
            return BadRequest(new { message = ex.Message });
        }
    }

    [HttpPost("{userId}/reset-password")]
    [Authorize(Policy = "RequireAdmin")]
    public async Task<ActionResult<User>> ResetUserPassword(long userId, [FromBody] PasswordResetConfirmRequest? request, [FromQuery] string? newPassword)
    {
        var email = GetUserEmail();
        var pwd = !string.IsNullOrWhiteSpace(request?.NewPassword) ? request.NewPassword : newPassword;
        if (string.IsNullOrWhiteSpace(pwd))
        {
            return BadRequest(new { message = "New password is required" });
        }

        try
        {
            var user = await _userService.ResetUserPasswordAsync(userId, pwd, email);
            return Ok(user);
        }
        catch (ArgumentException ex)
        {
            return BadRequest(new { message = ex.Message });
        }
        catch (InvalidOperationException ex)
        {
            return BadRequest(new { message = ex.Message });
        }
        catch (Exception ex)
        {
            return BadRequest(new { message = ex.Message });
        }
    }

    [HttpPost("transfer-ownership")]
    [Authorize(Policy = "RequireAdmin")]
    public async Task<ActionResult<Workspace>> TransferWorkspaceOwnership([FromQuery] long newOwnerId)
    {
        var email = GetUserEmail();
        try
        {
            var workspace = await _userService.TransferWorkspaceOwnershipAsync(newOwnerId, email);
            return Ok(workspace);
        }
        catch (ArgumentException ex)
        {
            return BadRequest(new { message = ex.Message });
        }
        catch (InvalidOperationException ex)
        {
            return BadRequest(new { message = ex.Message });
        }
    }

    [HttpPut("availability")]
    public async Task<ActionResult<User>> UpdateAvailabilityStatus([FromQuery] string status)
    {
        var email = GetUserEmail();
        try
        {
            var user = await _userService.UpdateAvailabilityStatusAsync(status, email);
            return Ok(user);
        }
        catch (ArgumentException ex)
        {
            return BadRequest(new { message = ex.Message });
        }
    }

    [HttpGet("productivity")]
    public async Task<ActionResult<List<TeamProductivityDto>>> GetTeamProductivity()
    {
        var email = GetUserEmail();
        try
        {
            var productivity = await _productivityService.GetTeamProductivityAsync(email);
            return Ok(productivity);
        }
        catch (Exception ex)
        {
            return BadRequest(new { message = ex.Message });
        }
    }

    private string GetUserEmail()
    {
        return User.FindFirstValue(ClaimTypes.Name) ?? User.FindFirstValue(ClaimTypes.Email) ?? string.Empty;
    }
}
